### PR TITLE
Dance AI: hot-swap live preview

### PR DESCRIPTION
**Pull request by sanchitmalhotra126**
Enable hot-swapping live preview by updating the preview effects without calling reset() for potential use in upcoming Dance AI modal UI changes. I also fixed a few lifecycle things to make sure blocks and previews were rendering smoothly.

Example (I overrode the "regenerate" button to produce a random AI output on click for this demo)

https://github.com/code-dot-org/code-dot-org/assets/85528507/f60afbbf-a7bb-4fe6-9e7b-a1a7dc9fdc63

## Testing story

Tested locally.

### CODE REVIEW

File: apps/src/dance/ai/AiBlockPreview.tsx [hunk 1]
```diff
@@ -18,7 +18,23 @@ const AiBlockPreview: React.FunctionComponent<AiBlockPreviewProps> = ({
 }) => {
   const blockPreviewContainerRef = useRef<HTMLSpanElement>(null);
   const refTimer = useRef<number | null>(null);
+  const workspaceRef = useRef<Workspace | null>(null);
 
+  // Create the workspace once the container has been rendered.
```
Comment (sanchitmalhotra126) [inline apps/src/dance/ai/AiBlockPreview.tsx:24]: @bencodeorg you had the [right idea](https://github.com/code-dot-org/code-dot-org/pull/54602#discussion_r1378004657) here :) turns out that if a workspace has no parent workspace, calling `dispose()` will not only remove the workspace from the DOM, but will also remove its parent element, because Blockly assumes it's the top level workspace. Not sure why exactly that's the case, but it became an issue here since calling `dispose()` would remove the `<span>` that the workspace was rendered into, so subsequent workspaces were being created with empty elements.

I updated this along the lines of what you suggested - creating a single workspace once the container has mounted, and switched to using `workspace.clear()` as the cleanup effect, which will just remove all blocks. We still call `workspace.dispose()` but only when the entire component is unmounting.

**Review by breville — APPROVED** [review]
Not super familiar with all of this code now, but this will be great to get in!  I did merge it into my WIP UI code and it looked good.  There was one blink as the live preview is reset as the UI modal re-renders due to other state changing, but I think we can fix-forward that issue once everything is merged.

File: apps/src/dance/ai/AiBlockPreview.tsx (hunk 1 above)
Comment (bencodeorg) [reply apps/src/dance/ai/AiBlockPreview.tsx:24]: Oh nice, good catch!

Comment (fisher-alice) [reply apps/src/dance/ai/AiBlockPreview.tsx:24]: Interesting! I was curious about the difference between `dispose` defined [here](https://github.com/google/blockly/blob/develop/core/workspace.ts#L154) and `clear`. It looks like `dispose` calls on `clear`, unregisters workspace in DB and reassigns `listeners.length`. Interesting that the parent element somehow gets deleted. I wonder if that's unintended.

File: apps/src/dance/ai/AiBlockPreview.tsx [hunk 2]
```diff
@@ -19,7 +19,23 @@ const AiBlockPreview: React.FunctionComponent<AiBlockPreviewProps> = ({
 }) => {
   const blockPreviewContainerRef = useRef<HTMLSpanElement>(null);
   const refTimer = useRef<number | null>(null);
+  const workspaceRef = useRef<Workspace | null>(null);
```
Comment (bencodeorg) [inline apps/src/dance/ai/AiBlockPreview.tsx:22]: Help me understand the use of a ref here (vs other hooks) -- I always think of refs as primarily used to reference some DOM node. React docs describe it as storing values "not needed for rendering" -- that's not totally true here, right (ie, we need it to render)?

**Review by bencodeorg — APPROVED** [review]
Very cool! One clarifying question about refs but this looks great.

File: apps/src/dance/ai/AiBlockPreview.tsx (hunk 2 above)
Comment (breville) [reply apps/src/dance/ai/AiBlockPreview.tsx:22]: I've started using `useRef` more, and currently understand that it gives us a variable that lives between re-renders, and that doesn't itself cause a re-render when changed.  Given that updating a variable created with `useState` triggers an immediate re-render (even before subsequent updates happen in the same function!), we want to be judicious in our use of `useState`, limiting it to state changes that we know will trigger an immediate re-render.  `useRef` then gives us a way to store additional variables, which can indeed be consumed by the rendering process, without actually triggering a re-render.

File: apps/src/dance/ai/AiBlockPreview.tsx (hunk 1 above)
Comment (sanchitmalhotra126) [reply apps/src/dance/ai/AiBlockPreview.tsx:24]: Yeah, looks like that happens in WorkspaceSvg [here](https://github.com/google/blockly/blob/9d117674ce64e88db35edd4426334fa4ae23d832/core/workspace_svg.ts#L887). Seems to be at least intentional but I'm curious what the rationale is

File: apps/src/dance/ai/AiBlockPreview.tsx (hunk 2 above)
Comment (sanchitmalhotra126) [reply apps/src/dance/ai/AiBlockPreview.tsx:22]: ^ yep exactly! In transitioning from class components to functional components with hooks, refs became more of a replacement for instance variables, rather than just DOM node references. In this case, I think the distinction is that the workspace object 1) isn't strictly a value that drives rendering in the React tree directly (as opposed to some simple state value like `shouldShowButton`) and 2) does render something on the page, but does so outside of the React virtual DOM (instead via direct DOM manipulation).

In other words - the workspace causes stuff to render, but not because of changes to workspace object itself, but instead by explicit calls from its/Blockly's API (`clear`, `dispose`, `newBlock`, etc)

Comment (bencodeorg) [reply apps/src/dance/ai/AiBlockPreview.tsx:22]: Thanks for explaining!
